compiler/rustc_const_eval/src/const_eval/eval_queries.rs

```diff
@@ -46,7 +47,11 @@ fn eval_body_using_ecx<'mir, 'tcx>(
     );
     let layout = ecx.layout_of(body.bound_return_ty().instantiate(tcx, cid.instance.args))?;
     assert!(layout.is_sized());
-    let ret = ecx.allocate(layout, MemoryKind::Stack)?;
+    let ret = if tcx.is_static(cid.instance.def_id()) && cid.promoted.is_none() {
```

Comment:
If you move up the `let intern_kind`, I think you can avoid duplicating the "is this truly a static" logic. In particular then the interner and the allocator here use the same variable to decide how to treat the root allocation. That seems preferable.